### PR TITLE
feat(server): Rate limit buckets by namespace

### PR DESCRIPTION
**Pull request by TBS1996**
part of: https://github.com/getsentry/relay/issues/2716

We want to also have global ratelimits based on namespace. This is done by grouping bucket-vectors by namespace and sending each group individually to check for ratelimiting.

### CODE REVIEW

File: relay-quotas/src/redis.rs
Comment (TBS1996) [inline relay-quotas/src/redis.rs:132]: no longer necessary because we construct the redis key separately for global scoped quotas.

File: tests/integration/test_outcome.py
Comment (TBS1996) [inline tests/integration/test_outcome.py]: the '-1' is a problem I'm thinking of a solution for now.  
 
When multiple quotas matches the same item, like here in the case of both a namespace quota and one without, we count down when it matches the quota without namespace even though it gets ratelimited by the namespace quota right after.

Comment (TBS1996) [reply tests/integration/test_outcome.py]: The ugly immediate fix is like this:

``` Rust
        global_quotas.sort_by_key(|x| x.namespace.is_none());
        for quota in global_quotas {
            let is_rate_limited = self
                .global_limits
                .is_rate_limited(&mut client, &quota, quantity)
                .map_err(RateLimitingError::Redis)?;

            if is_rate_limited {
                let retry_after = self.retry_after((quota.expiry() - timestamp).as_secs());
                rate_limits.add(RateLimit::from_quota(&quota, &item_scoping, retry_after));
                break;
            }
        }
```

it works by ensuring the more restrictive global quotas come first, this won't scale though.  
 
working on something more generic now

Comment (TBS1996) [reply tests/integration/test_outcome.py]: done.

File: Cargo.lock
Comment (TBS1996) [inline Cargo.lock:3856]: need it for group_by in filter_ratelimited

File: relay-metrics/src/bucket.rs
Comment (jjbayer) [inline relay-metrics/src/bucket.rs]: nit: If a function has to do work, it's good to reflect it in the name.

```suggestion
    pub fn parse_namespace(&self) -> Result<MetricNamespace, ParseMetricError> {
```

File: relay-quotas/Cargo.toml
Comment (jjbayer) [inline relay-quotas/Cargo.toml:25]: TODO: necessary?

File: relay-quotas/src/global.rs
Comment (jjbayer) [inline relay-quotas/src/global.rs:73]: I would make the namespace part of `prefix` here. The global quota logic does not need to know _what_ it is limiting, as long as it has a unique prefix to keep quotas apart.

File: relay-quotas/src/global.rs
Comment (jjbayer) [inline relay-quotas/src/global.rs]:
```suggestion
    pub fn filter_rate_limited<'a>(
```

File: relay-quotas/src/global.rs
Comment (jjbayer) [inline relay-quotas/src/global.rs]: nit: I like having constructor methods (`new`) at the beginning of the `impl` block.

File: relay-quotas/src/quota.rs
Comment (jjbayer) [inline relay-quotas/src/quota.rs:359]: I guess `insta` did this, but I remember the beta linter complained about triple `###` a while back -> check if `cargo + nightly clippy` likes this.

File: relay-quotas/src/quota.rs
Comment (jjbayer) [inline relay-quotas/src/quota.rs:58]: I dislike adding the metrics namespace here, but I don't have a better idea at the moment.

File: relay-quotas/src/redis.rs
Comment (jjbayer) [reply relay-quotas/src/redis.rs:132]: See comment about making namespace part of the prefix. With that, I think this code can remain as it is.

**Review by jjbayer — COMMENTED** [review]
Overall looks good to me, see comment about putting namespace in the prefix.

File: relay-quotas/Cargo.toml
Comment (TBS1996) [reply relay-quotas/Cargo.toml:25]: you mean if the group_by thing is necessary?

File: relay-quotas/src/global.rs
Comment (TBS1996) [reply relay-quotas/src/global.rs:73]: is it worth the extra string allocation though? We only use `Key` when we put a new key into the hashmap, otherwise we use keyref which is a cheap copy type. If we change the 'Key' here we'd have to do away with the cheap keyref type.

File: relay-quotas/src/quota.rs
Comment (TBS1996) [reply relay-quotas/src/quota.rs:359]: i tried, it's fine

File: relay-quotas/src/redis.rs
Comment (TBS1996) [reply relay-quotas/src/redis.rs:132]: left comment about prefix below

File: relay-quotas/src/quota.rs
Comment (TBS1996) [reply relay-quotas/src/quota.rs:58]: yeah I have the same opinion

File: CHANGELOG.md
Comment (olksdr) [inline CHANGELOG.md]: Please, move this to `Unreleased` section

File: CHANGELOG.md
Comment (jjbayer) [inline CHANGELOG.md]:
```suggestion
- Add ability to rate limit metric buckets by namespace. ([#2941](https://github.com/getsentry/relay/pull/2941))
```

File: relay-quotas/Cargo.toml
Comment (jjbayer) [reply relay-quotas/Cargo.toml:25]: Yeah I was wondering whether we could work without the dependency. But it does improve readability, so it's fine with me.

File: relay-server/src/services/processor.rs
Comment (jjbayer) [inline relay-server/src/services/processor.rs:1715]: We could probably prevent moving buckets around here, by working with a hashmap of references and only recreating the vector at the end. Not a blocker for me though.

**Review by jjbayer — APPROVED** [review]
@TBS1996 there's still at least one open comment, please check before merging.

Summary of pair review:

* Sprinkling `namespace` fields across types is not great because it introduces a datatype-specific attribute into a general type. The best approach would be to have a generic matching language for quota, similar to dynamic sampling, but that's out of scope for this PR.
* We discarded the possibility of a new `QuotaScope` `"global-namespace"` and encoding the namespace into the `scope_id`. This would allow us to reuse a lot of code, but it loses the flexibility of configuring per-org or per-project namespace limits.
* We can easily get rid of the namespace field(s) in the future if we want to reconsider, as long as we make sentry stop sending namespace-limits as well.

File: relay-server/src/services/processor.rs
Comment (TBS1996) [reply relay-server/src/services/processor.rs:1715]: gave it one try and the `BucketsView` taking in a `&[Bucket]` rather than a `&[&Buckets]` made it harder. Can probably make it work but don't wanna block this PR any longer atm

**Pull request merged**
